src/attributes/classes.js

```diff
@@ -100,60 +99,29 @@ jQuery.fn.extend( {
 	},
 
 	toggleClass: function( value, stateVal ) {
-		var type = typeof value;
-
-		if ( typeof stateVal === "boolean" && type === "string" ) {
-			return stateVal ? this.addClass( value ) : this.removeClass( value );
-		}
-
-		if ( jQuery.isFunction( value ) ) {
-			return this.each( function( i ) {
-				jQuery( this ).toggleClass(
-					value.call( this, i, getClass( this ), stateVal ),
-					stateVal
-				);
-			} );
-		}
-
-		return this.each( function() {
-			var className, i, self, classNames;
-
-			if ( type === "string" ) {
-
-				// Toggle individual class names
-				i = 0;
-				self = jQuery( this );
-				classNames = value.match( rnotwhite ) || [];
-
-				while ( ( className = classNames[ i++ ] ) ) {
-
-					// Check each className given, space separated list
-					if ( self.hasClass( className ) ) {
-						self.removeClass( className );
-					} else {
-						self.addClass( className );
-					}
-				}
-
-			// Toggle whole class name
-			} else if ( value === undefined || type === "boolean" ) {
-				className = getClass( this );
-				if ( className ) {
+		var type = typeof value,
+			classNames = type === "string" ? value.match( rnotwhite ) : "",
+			checker = typeof stateVal === "boolean" ?
+				function() { return !stateVal; } :
+				jQuery.fn.hasClass;
+
+		return this.each( function( i ) {
+			var className,
+				self = jQuery( this ),
+				c = 0;
+
+			if ( type === "function" ) {
+				classNames = value.call( this, i, getClass( this ), stateVal )
+					.match( rnotwhite ) || [];
+			}
 
-					// Store className if set
-					dataPriv.set( this, "__className__", className );
-				}
+			// Toggle individual class names based on presence or stateVal
+			while ( ( className = classNames[ c++ ] ) ) {
 
-				// If the element has a class name or if we're passed `false`,
-				// then remove the whole classname (if there was one, the above saved it).
-				// Otherwise bring back whatever was previously saved (if anything),
-				// falling back to the empty string if nothing was stored.
-				if ( this.setAttribute ) {
-					this.setAttribute( "class",
-						className || value === false ?
-						"" :
-						dataPriv.get( this, "__className__" ) || ""
-					);
+				if ( checker.call( self, className ) ) {
```

Comment:
Ah, right, we'd need two checks against `stateVal`. It might still be preferable, but I'm now less enthusiastic.
```js
if ( stateVal !== true && self.hasClass( className ) ) {
	self.removeClass( className );
} else if ( stateVal !== false ) {
	self.addClass( className )
}
```